middleware: Remove hard dependency on Gorilla Sessions

lstoll opened the pull request:
Right now using the middleware is hard bound to using gorilla's sessions library, which makes it annoying to use in any project that isn't already using that for sessions. This adds a more generic load/save implementation for a custom type, that consumers can implement with whatever their preferred session management scheme is.

A implementation of this Interface is provided for gorilla sessions, and it has been added in a way that is fully backwards compatible - existing code needs no changes. That has been marked as deprecated though, to indicate that we'll want to get rid of it at some point.

The existing tests exercise both the interface and the adapter.

lstoll committed:
middleware: Remove hard dependency on Gorilla Sessions

Right now using the middleware is hard bound to using gorilla's sessions
library, which makes it annoying to use in any project that isn't already using
that for sessions. This adds a more generic load/save implementation for a
custom type, that consumers can implement with whatever their preferred session
management scheme is.

A implementation of this Interface is provided for gorilla sessions, and it has
been added in a way that is fully backwards compatible - existing code needs no
changes. That has been marked as deprecated though, to indicate that we'll want
to get rid of it at some point.

The existing tests exercise both the interface and the adapter.

sr commented on middleware/middleware.go:
```diff
@@ -15,14 +16,29 @@ import (
+type Sessions interface {
+	GetSession(*http.Request) (*SessionData, error)
+	SaveSession(http.ResponseWriter, *http.Request, *SessionData) error
+}
```
nit, but WDYT of `type SessionStoreV2` with `Get` and `Put`? Less stutter-y. More generally I wonder if it's even worth maintaining backward compatibility in lstoll/oidc?

lstoll replied:
Would you change the value on the `Middleware` struct to `SessionStoreV2` as well?

Also I waffle a lot on the stutter thing on interfaces. Really generic stuff like `Get` often conflicts on implementations if you're passing something that does multiple things, I don't like having to have a sub-struct to work around that. But yeah, stutter.

sr replied:
> Would you change the value on the Middleware struct to SessionStoreV2 as well?

Probably, yes.

> Really generic stuff like Get often conflicts on implementations 

I mean things like the Reader interface is also super generic, and isn't a problem in practice I dunno. https://pkg.go.dev/golang.org/x/crypto/acme/autocert#DirCache also.

lstoll replied:
Fiiiiiine

lstoll replied:
Dropped backwards compatibility, renamed things

sr reviewed (approved):
rad

lstoll committed:
Note that they should be secret

lstoll committed:
rename as suggested

lstoll committed:
Better when I save the file

lstoll committed:
Merge remote-tracking branch 'origin/main' into lstoll-gorilla-soft-dep

lstoll committed:
Drop backwards compat/gorilla shim

Not worth carrying it for longer. Just drop use of gorilla, stub in a basic
memory implementation for testing.

lstoll committed:
sp

sr commented on middleware/memory_session.go:
```diff
@@ -0,0 +1,101 @@
+)
+
+// MemorySessionStore is a simple session store, that tracks state in memory. It
+// is mainly used for testing, it is not suitible for anything outside a single
```
```suggestion
// is mainly used for testing, it is not suitable for anything outside a single
```

sr commented on middleware/memory_session.go:
```diff
@@ -0,0 +1,101 @@
+	return nil
+}
+
+func (m *MemorySessionStore) init() error {
```
`NewMemoryStore(*http.Cookie) SessionStore` perhaps

lstoll replied:
Did that. Make the struct private too, not sure what having it public adds when it returns an interface.

lstoll committed:
Add a constructor

lstoll committed:
sp
